### PR TITLE
A few fixes, including NRE & Object Refs

### PR DESCRIPTION
**Pull request by AlienXAXS**
- Should fix "Object reference not set to the instance of an object"
- Made the downloader actually properly multi-threaded now, it downloads from multiple servers, not just one
- Fixed an error in the GetNextAvailableServerEntry method that would sometimes throw a NRE

### CODE REVIEW

File: RXPatchLib/UpdateServerSelector.cs [hunk 1]
```diff
@@ -35,18 +35,44 @@ public class UpdateServerSelector
...
+                    // Order them by connection count and take the top one off the pile
+                    var selectedServer = selectedServers.OrderBy(x => x.ConnectionCount).DefaultIfEmpty(null).FirstOrDefault();
+
+                    // If we didnt get null, ++ connection count and return it, otherwise return null
+                    if (selectedServer != null)
+                    {
+                        selectedServer.ConnectionCount++;
+
+                        RxLogger.Logger.Instance.Write(
+                            $"I have picked the server {selectedServer.UpdateServer.Uri.AbsoluteUri} as it has only {selectedServer.ConnectionCount} connections agaist it");
```
Comment (Cronus89) [inline RXPatchLib/UpdateServerSelector.cs]: against*

File: RXPatchLib/WebPatchSource.cs [hunk 2]
```diff
@@ -209,12 +200,9 @@ private void UnlockDownload()
                     }
                     catch (TooManyRetriesException)
                     {
+                        if (thisPatchServer != null) thisPatchServer.HasErrored = true;
                         AXDebug.AxDebuggerHandler.Instance.RemoveDownload(guid);
 
-                        // Mark current mirror failed
```
Comment (JAJames) [inline RXPatchLib/WebPatchSource.cs]: What's the purpose of this change?

File: RXPatchLib/WebPatchSource.cs [hunk 3]
```diff
@@ -232,10 +220,6 @@ private void UnlockDownload()
                         AXDebug.AxDebuggerHandler.Instance.RemoveDownload(guid);
                         RxLogger.Logger.Instance.Write($"Invalid file hash for {subPath} - Expected hash {hash}, requeuing download");
 
-                        // Mark current mirror failed
```
Comment (JAJames) [inline RXPatchLib/WebPatchSource.cs]: If a mirror has invalid data, we should continue to mark it failed.

File: RXPatchLib/WebPatchSource.cs [hunk 4]
```diff
@@ -193,14 +192,6 @@ private void UnlockDownload()
                                     $"Error while attempting to transfer the file.\r\n{e.Message}\r\n{e.StackTrace}");
                                 cancellationToken.ThrowIfCancellationRequested();
                                 AXDebug.AxDebuggerHandler.Instance.RemoveDownload(guid);
-
```
Comment (JAJames) [inline RXPatchLib/WebPatchSource.cs]: If we're receiving 400/client errors from the server, they're not going to be resolvable and we should continue fail fast here.

File: RXPatchLib/WebPatchSource.cs [hunk 5]
```diff
@@ -100,6 +100,10 @@ private void UnlockDownload()
         {
             string filePath = GetSystemPath(subPath);
             var guid = Guid.NewGuid();
+            var thisPatchServer = _patcher.UpdateServerSelector.GetNextAvailableServerEntry();
```
Comment (JAJames) [inline RXPatchLib/WebPatchSource.cs]: We shouldn't move this up here, because if we do then failover will no longer function because a new mirror will never be selected.

File: RXPatchLib/UpdateServerSelector.cs [hunk 6]
```diff
@@ -35,18 +35,44 @@ public class UpdateServerSelector
         /// <returns></returns>
         public UpdateServerEntry GetNextAvailableServerEntry()
         {
-            if (CurrentHostsList.Count == 0)
-                return null;
+            var ServerRecordsToTake = 4;
```
Comment (JAJames) [inline RXPatchLib/UpdateServerSelector.cs]: Can we make a global constant somewhere for this?

File: RXPatchLib/WebPatchSource.cs (hunk 5 above)
Comment (AlienXAXS) [reply RXPatchLib/WebPatchSource.cs]: Doesn't really matter that it's up there, can be fixed by asking for a new host during the catch.

**Comment by AlienXAXS** [comment]
Ready for another review.
